### PR TITLE
Add collaboration service

### PR DESCRIPTION
**Pull request by chownces**

### CODE REVIEW

File: backend/apps/collaboration/knexfile.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+const knexConfig = {
+  client: 'pg',
+  connection: {
+    host: process.env.COLLABORATION_SERVICE_SQL_DATABASE_HOST,
+    port: process.env.COLLABORATION_SERVICE_SQL_DATABASE_PORT,
+    user: process.env.COLLABORATION_SERVICE_SQL_DATABASE_USER,
+    password: process.env.COLLABORATION_SERVICE_SQL_DATABASE_PASSWORD,
+    database: process.env.COLLABORATION_SERVICE_SQL_DATABASE_NAME,
+  },
```
Comment (chownces) [inline backend/apps/collaboration/knexfile.ts]: Abstract this out via getDbConfForSvc

File: backend/apps/collaboration/src/collaboration.service.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    private readonly sessionDaoService: SessionDaoService,
+  ) {}
+
+  createCollabSession(createSessionInfo: CreateSessionInfo) {
+    const graphInfo = {
+      ...createSessionInfo,
+      userIds: createSessionInfo.userIds.map((userId) => ({ userId })),
+    };
+    return this.sessionDaoService.create(graphInfo);
+  }
```
Comment (chownces) [inline backend/apps/collaboration/src/collaboration.service.ts]: Add additional validation checks for the methods in this service layer

File: backend/apps/collaboration/src/config/configuration.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  const userServiceOptions = getRmqOptionsForQueue(RmqQueue.USER);
+  const questionServiceOptions = getRmqOptionsForQueue(RmqQueue.QUESTION);
+
+  const databaseConfigurationOptions: DatabaseConfigurationOptions = {
+    host: process.env.COLLABORATION_SERVICE_SQL_DATABASE_HOST,
+    port: process.env.COLLABORATION_SERVICE_SQL_DATABASE_PORT,
+    user: process.env.COLLABORATION_SERVICE_SQL_DATABASE_USER,
+    password: process.env.COLLABORATION_SERVICE_SQL_DATABASE_PASSWORD,
+    database: process.env.COLLABORATION_SERVICE_SQL_DATABASE_NAME,
+  };
```
Comment (chownces) [inline backend/apps/collaboration/src/config/configuration.ts]: Abstract this out with getDbConfForSvc

File: backend/apps/gateway/src/app.module.ts [hunk 4]
```diff
@@ -9,7 +9,27 @@ import { AuthController } from './controllers/auth.controller';
...
+  microservice: string,
+  optionsKey: string,
+): Provider => ({
+  provide: microservice,
+  useFactory: (configService: ConfigService) => {
+    const microserviceOptions = configService.get(optionsKey);
+    return ClientProxyFactory.create(microserviceOptions);
+  },
+  inject: [ConfigService],
+});
```
Comment (chownces) [inline backend/apps/gateway/src/app.module.ts]: Remove this in favour of the one in common lib now

File: backend/apps/gateway/src/websocket-gateways/yjs.gateway.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      docName,
+    });
+
+    const mdb = new MongodbPersistence(
+      'mongodb://localhost:27017/peer-prep-collaboration',
+      {
+        flushSize: 100,
+        multipleCollections: true,
+      },
+    );
```
Comment (chownces) [inline backend/apps/gateway/src/websocket-gateways/yjs.gateway.ts:63]: Let's abstract this in a future PR when dockerizing mongo

File: backend/apps/gateway/src/app.module.ts [hunk 6]
```diff
@@ -18,27 +38,14 @@ import { Service } from '@app/interservice-api/services';
...
-    },
-    {
-      provide: Service.USER_SERVICE,
-      useFactory: (configService: ConfigService) => {
-        const userServiceOptions = configService.get('userServiceOptions');
-        return ClientProxyFactory.create(userServiceOptions);
-      },
-      inject: [ConfigService],
-    },
+    YjsGateway,
```
Comment (chownces) [inline backend/apps/gateway/src/app.module.ts:32]: Let's explore shifting the websocket connection to a separate gateway in the future

File: backend/apps/user/src/auth/auth.controller.ts [hunk 7]
```diff
@@ -27,4 +30,15 @@ export class AuthController {
...
   }
+
+  @MessagePattern(UserServiceApi.GENERATE_WEBSOCKET_TICKET)
+  generateWebsocketTicket(createTicketInfo: CreateWebsocketTicketInfo) {
+    return this.authService.generateWebsocketTicket(createTicketInfo);
+  }
+
+  @MessagePattern(UserServiceApi.CONSUME_WEBSOCKET_TICKET)
+  consumeWebsocketTicket(ticketId: string) {
+    console.log('calling auth consume controoler');
```
Comment (chownces) [inline backend/apps/user/src/auth/auth.controller.ts]: Remove console.log

**Review by aikenwx — APPROVED** [review]
LGTM

**Pull request merged**